circuit-benchmarks/src/bytecode_circuit.rs

```diff
@@ -0,0 +1,251 @@
+        );
+
+        let num_rows = 1 << degree;
+        const NUM_BLINDING_ROWS: usize = 7 - 1;
+        let instance = vec![randomness; num_rows - NUM_BLINDING_ROWS];
+
+        // Initialize the polynomial commitment parameters
+        let mut rng = XorShiftRng::from_seed([
+            0x59, 0x62, 0xbe, 0x5d, 0x76, 0x3d, 0x31, 0x8d, 0x17, 0xdb, 0x37, 0x32, 0x54, 0x06,
+            0xbc, 0xe5,
+        ]);
+
+        // Bench setup generation
+        let setup_message = format!("Setup generation with degree = {}", degree);
+        let start1 = start_timer!(|| setup_message);
+        let general_params = ParamsKZG::<Bn256>::setup(degree, &mut rng);
+        let verifier_params: ParamsVerifierKZG<Bn256> = general_params.verifier_params().clone();
+        end_timer!(start1);
+
+        // Initialize the proving key
+        let vk = keygen_vk(&general_params, &bytecode_circuit).expect("keygen_vk should not fail");
+        let pk =
+            keygen_pk(&general_params, vk, &bytecode_circuit).expect("keygen_pk should not fail");
+        // Create a proof
+        let mut transcript = Blake2bWrite::<_, G1Affine, Challenge255<_>>::init(vec![]);
+
+        // Bench proof generation time
+        let proof_message = format!("Bytecode Proof generation with {} rows", degree);
+        let start2 = start_timer!(|| proof_message);
+        create_proof::<
+            KZGCommitmentScheme<Bn256>,
+            ProverSHPLONK<'_, Bn256>,
+            Challenge255<G1Affine>,
+            XorShiftRng,
+            Blake2bWrite<Vec<u8>, G1Affine, Challenge255<G1Affine>>,
+            BytecodeCircuitTester<Fr>,
+        >(
+            &general_params,
+            &pk,
+            &[bytecode_circuit],
+            &[&[instance.as_slice()][..]][..],
+            rng,
+            &mut transcript,
+        )
+        .expect("proof generation should not fail");
+        let proof = transcript.finalize();
+        end_timer!(start2);
+
+        // Bench verification time
+        let start3 = start_timer!(|| "Bytecode Proof verification");
+        let mut verifier_transcript = Blake2bRead::<_, G1Affine, Challenge255<_>>::init(&proof[..]);
+        let strategy = SingleStrategy::new(&general_params);
+
+        verify_proof::<
+            KZGCommitmentScheme<Bn256>,
+            VerifierSHPLONK<'_, Bn256>,
+            Challenge255<G1Affine>,
+            Blake2bRead<&[u8], G1Affine, Challenge255<G1Affine>>,
+            SingleStrategy<'_, Bn256>,
+        >(
+            &verifier_params,
+            pk.get_vk(),
+            strategy,
+            &[&[instance.as_slice()][..]][..],
+            &mut verifier_transcript,
+        )
+        .expect("failed to verify bench circuit");
+        end_timer!(start3);
+    }
+
+    fn generate_code_bytes<F: Field>(
+        bytecodes_len: u32,
+        randomness: F,
+    ) -> Vec<UnrolledBytecode<F>> {
+        let sample_hexbytes: &str = "608060405260405162000f6238038062000f6283398101604081905262000026\
```

Comment:
Why not generate some bytecode here instead? Seems cleaner and makes it more clear what is being used for testing. The circuit should accept all reasonable inputs so shouldn't be hard to make on the fly.

Reply:
You mean use bytecode! marco to make the code readable? I thought it is hard to write a long reasonable bytecode array when I wrote the test, meanwhile such long bytecode array also makes a long file (this snippet has 4000 opcodes), so I chose to copy a compiled bytes from etherscan.io. 
I think it's ok to translate them back to op code array (if a long file is ok), or I can just a relative smaller one with more repeat times (to max fill the circuit table up).

Reply:
The opcodes should have negligible impact on the performance so doesn't really matter if it's "real" bytecode or not. So just filling the bytes looping over all the different opcode values until the circuit is full should be more than fine.

Reply:
got it, fixed by traversal the opcode byte values.